sycl/include/sycl/builtins.hpp

```diff
@@ -2073,19 +2112,22 @@ detail::enable_if_t<detail::is_gentype<T>::value, T> bitselect(T a, T b,
 template <typename T>
 detail::enable_if_t<detail::is_sgentype<T>::value, T> select(T a, T b,
                                                              bool c) __NOEXC {
-  return __sycl_std::__invoke_select<T>(a, b, static_cast<int>(c));
-}
-
-// mgentype select (mgentype a, mgentype b, marray<bool, { N }> c)
-template <typename T,
-          typename = std::enable_if_t<detail::is_mgenfloat<T>::value>>
-sycl::marray<detail::marray_element_t<T>, T::size()>
-select(T a, T b, sycl::marray<bool, T::size()> c) __NOEXC {
-  sycl::marray<detail::marray_element_t<T>, T::size()> res;
-  for (int i = 0; i < a.size(); i++) {
-    res[i] = select(a[i], b[i], c[i]);
-  }
-  return res;
+  if constexpr (std::is_same_v<T, double> || std::is_same_v<T, long> ||
+                std::is_same_v<T, unsigned long>)
+    return __sycl_std::__invoke_select<T>(a, b, static_cast<long>(c));
+  else if constexpr (std::is_same_v<T, char> ||
+                     std::is_same_v<T, signed char> ||
+                     std::is_same_v<T, unsigned char>)
+    return __sycl_std::__invoke_select<T>(a, b, static_cast<char>(c));
+  else if constexpr (std::is_same_v<T, short> ||
+                     std::is_same_v<T, unsigned short> ||
+                     std::is_same_v<T, half>)
+    return __sycl_std::__invoke_select<T>(a, b, static_cast<short>(c));
+  else if constexpr (std::is_same_v<T, long long> ||
+                     std::is_same_v<T, unsigned long long>)
+    return __sycl_std::__invoke_select<T>(a, b, static_cast<long long>(c));
+  else
+    return __sycl_std::__invoke_select<T>(a, b, static_cast<int>(c));
```

Comment:
Should we do `sizeof(...)`-based checks instead?

Reply:
I thought this is not reliable based on Alexey's comment, so I closed it. Should we do sizeof-based checks in such case?

Reply:
Misunderstood this, sorry, implemented.